### DISCUSSION TITLE
Question about custom session objects

### DISCUSSION THREAD

**Discussion by randomprogramming** [nextauthjs/next-auth#4223] [Help] [answered]
### Question 💬

Hello this is my `[...nextauth].ts` file:

```
import NextAuth from "next-auth";
import DiscordProvider from "next-auth/providers/discord";
import { PrismaAdapter } from "@next-auth/prisma-adapter";
import { PrismaClient } from "@prisma/client";

export default NextAuth({
  adapter: PrismaAdapter(new PrismaClient()),
  providers: [
    DiscordProvider({
      clientId: process.env.DISCORD_ID,
      clientSecret: process.env.DISCORD_SECRET,
    }),
  ],
  callbacks: {
    session: ({ session, user }) => {
      console.log("USER:");
      console.log(user);

      const newSession = {
        ...session,
        user: { roles: user.roles, ...session.user },
      };

      console.log("Returning new Session:");
      console.log(newSession);

      return newSession;
    },
  },
});
```

I needed a way to conditionally render some frontend elements based on the role of the user, but I didn't want to create a separate call to get the user roles from the database, so I ended up with this. The above code prints:

```
USER:
{
  id: 'a9d0afdf-6aeb-4927-a14d-bdbef8a33ef7',
  name: 'MyDiscordName',
  email: 'MyDiscordEmail@gmail.com',
  emailVerified: null,
  image: 'MyDiscordImage',
  roles: [ 'admin', 'tester' ]
}
Returning new Session:
{
  user: {
    roles: [ 'admin', 'tester' ],
    name: 'MyDiscordName',
    email: 'MyDiscordEmail@gmail.com',
    image: 'MyDiscordImage'
  },
  expires: '2022-04-18T13:03:27.072Z'
}
```

The user schema for Prisma looks like this:
```
model User {
  id            String    @id @default(uuid())
  name          String?
  email         String?   @unique
  emailVerified DateTime?
  image         String?
  roles         String[]
  accounts      Account[]
  sessions      Session[]
}
```

Now I have two questions:
1. Is this the proper way of doing this? I literally couldn't find anything about this in the docs, and there really isn't much information on the internet about this scenario. I only saw examples where people used the JWT callback, but that obviously won't work for me because of the adapter.
2. Is there any way to type the `user` parameter in the `session` callback? In the current setup, typescript evaluates `user.roles` as `unknown` which isn't true..

### How to reproduce ☕️

Entire code is above.

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by balazsorban44** [comment, marked as answer]
1. Yes, https://next-auth.js.org/configuration/callbacks#session-callback
2. https://next-auth.js.org/getting-started/typescript

**Reply by randomprogramming** [reply]
Thank you :)

**Reply by Standaa** [reply]
@balazsorban44 How are we supposed to manage the same situation with a property that can be changed?
Let me explain: Say I want to fetch a user's phone number to display it in his settings. My user can then update it.
If I am not mistaken, NextAuth only allows attaching this information when a user logs in (when the session callback gets called).
If I want this information to be updated in the UI, I would have to log my user out after any profile update.
This would be quite clunky.
Is there another way to achieve the flow I just described?

**Reply by dhpancor** [reply]
> @balazsorban44 How are we supposed to manage the same situation with a property that can be changed? Let me explain: Say I want to fetch a user's phone number to display it in his settings. My user can then update it. If I am not mistaken, NextAuth only allows attaching this information when a user logs in (when the session callback gets called). If I want this information to be updated in the UI, I would have to log my user out after any profile update. This would be quite clunky. Is there another way to achieve the flow I just described?

I have the exact same problem, but with roles. What if the role of an user changed? I can't get Next-Auth to refetch the values for the account/user stored in the database once the user signs in.

**Reply by galer7** [reply]
@Standaa @dhpancor This might help: https://github.com/nextauthjs/next-auth/issues/2269#issuecomment-933506368
The links in the docs (advanced initialization): https://next-auth.js.org/configuration/initialization#advanced-initialization
I've just discovered this after some searching. I need this behavior as well when a user joins a group, and having that groupId on the session object would make my life easier.
